From af5a593e2b191c5f3cb6714d9b29599dab5dbc42 Mon Sep 17 00:00:00 2001
From: Tong Tiangen <tongtiangen@huawei.com>
Date: Wed, 17 Aug 2022 10:33:56 +0000
Subject: [PATCH 005/261] [Huawei] arm64/__mc_ex_table: fix __mc_ex_table
 do_sort() issue

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187496

--------------------------------

If ARCH_HAS_MC_EXTABLE is not set then the _mc_ex_table is empty.

There is a logic error in do_sort(). _mc_ex_table is different from
_ex_table. _ex_table is always not empty, while _mc_ex_table may be empty.
If _mc_ex_table is empty, there is a issue with setting the initial value
of mc_extable_index to 0.

Fixes: 7c3d926bb083 ("[Huawei] arm64: extable: add new extable type "__mc_ex_table"")
Signed-off-by: Tong Tiangen <tongtiangen@huawei.com>
---
 scripts/sorttable.h | 20 +++++++++-----------
 1 file changed, 9 insertions(+), 11 deletions(-)

diff --git a/scripts/sorttable.h b/scripts/sorttable.h
index 874cbd7046b0..d67242f8c75f 100644
--- a/scripts/sorttable.h
+++ b/scripts/sorttable.h
@@ -227,7 +227,6 @@ static int do_sort(Elf_Ehdr *ehdr,
 	Elf_Rel *mc_relocs = NULL;
 	int mc_relocs_size = 0;
 	char *mc_extab_image = NULL;
-	int mc_extab_index = 0;
 
 	shstrndx = r2(&ehdr->e_shstrndx);
 	if (shstrndx == SHN_XINDEX)
@@ -245,11 +244,6 @@ static int do_sort(Elf_Ehdr *ehdr,
 			extab_index = i;
 		}
 
-		if (!strcmp(secstrings + idx, "__mc_ex_table")) {
-			mc_extab_sec = s;
-			mc_extab_index = i;
-		}
-
 		if (!strcmp(secstrings + idx, ".symtab"))
 			symtab_sec = s;
 		if (!strcmp(secstrings + idx, ".strtab"))
@@ -262,11 +256,15 @@ static int do_sort(Elf_Ehdr *ehdr,
 			relocs_size = _r(&s->sh_size);
 		}
 
-		if ((r(&s->sh_type) == SHT_REL ||
-		     r(&s->sh_type) == SHT_RELA) &&
-		    r(&s->sh_info) == mc_extab_index) {
-			mc_relocs = (void *)ehdr + _r(&s->sh_offset);
-			mc_relocs_size = _r(&s->sh_size);
+		if (!strcmp(secstrings + idx, "__mc_ex_table")) {
+			mc_extab_sec = s;
+
+			if ((r(&s->sh_type) == SHT_REL ||
+				r(&s->sh_type) == SHT_RELA) &&
+				r(&s->sh_info) == i) {
+				mc_relocs = (void *)ehdr + _r(&s->sh_offset);
+				mc_relocs_size = _r(&s->sh_size);
+			}
 		}
 
 		if (r(&s->sh_type) == SHT_SYMTAB_SHNDX)
-- 
2.17.1

